bundle/edu.gemini.catalog/src/main/scala/edu/gemini/catalog/votable/VoTableParser.scala

```diff
@@ -220,36 +286,37 @@ trait VoTableParser {
         }.toMap
       // Filter magnitudes as a whole
-      val magnitudes = magFilter.filterMagnitudeFields(mags)
+      val magnitudes = adapter.filterAndDeduplicateMagnitudes(mags)
       // Link magnitudes with their errors
-      magnitudes.map(i => i.copy(error = magErrors.get(i.band))).filter(magFilter.validMagnitude)
+      magnitudes.map(i => i.copy(error = magErrors.get(i.band)))
     }
 
     def toSiderealTarget(id: String, ra: String, dec: String, pm: (Option[String], Option[String])): \/[CatalogProblem, SiderealTarget] = {
       for {
-        magFilter      <- magnitudesFilter
+        adapter        <- catalogAdapter
         r              <- Angle.parseDegrees(ra).leftMap(_ => FieldValueProblem(VoTableParser.UCD_RA, ra))
         d              <- Angle.parseDegrees(dec).leftMap(_ => FieldValueProblem(VoTableParser.UCD_DEC, dec))
         declination    <- Declination.fromAngle(d) \/> FieldValueProblem(VoTableParser.UCD_DEC, dec)
         properMotion   <- parseProperMotion(pm)
-        mags            = entries.filter(magnitudeField(_, magFilter))
-        magErrs         = entries.filter(magnitudeErrorField(_, magFilter))
-        magnitudes     <- mags.map(parseBands(magFilter)).toList.sequenceU
-        magnitudeErrs  <- magErrs.map(parseBands(magFilter)).toList.sequenceU
+        mags            = entries.filter(adapter.isMagnitudeField)
+        magErrs         = entries.filter(adapter.isMagnitudeErrorField)
+        magnitudes     <- mags.map(adapter.parseMagnitude).toList.sequenceU
+        magnitudeErrs  <- magErrs.map(adapter.parseMagnitude).toList.sequenceU
       } yield {
         val coordinates = Coordinates(RightAscension.fromAngle(r), declination)
-        val combMags    = combineWithErrorsAndFilter(magnitudes, magnitudeErrs, magFilter).sorted(VoTableParser.MagnitudeOrdering)
+        val combMags    = combineWithErrorsAndFilter(magnitudes, magnitudeErrs, adapter).sorted(VoTableParser.MagnitudeOrdering)
         SiderealTarget(id, coordinates, properMotion, combMags, None)
       }
     }
 
     val result = for {
-        id             <- entriesByUcd.get(VoTableParser.UCD_OBJID)
-        ra             <- entriesByUcd.get(VoTableParser.UCD_RA)
-        dec            <- entriesByUcd.get(VoTableParser.UCD_DEC)
-        (pmRa, pmDec)   = (entriesByUcd.get(VoTableParser.UCD_PMRA), entriesByUcd.get(VoTableParser.UCD_PMDEC))
-      } yield toSiderealTarget(id, ra, dec, (pmRa, pmDec))
-
-    result.getOrElse(\/.left(MissingValues(missing)))
+      adapter <- catalogAdapter
+      id      <- entries.get(adapter.idField) \/> MissingValue(adapter.idField)
+      ra      <- entries.get(adapter.raField) \/> MissingValue(adapter.raField)
+      dec     <- entries.get(adapter.decField) \/> MissingValue(adapter.decField)
+      (pmRa, pmDec)   = (entries.get(adapter.pmRaField), entries.get(adapter.pmDecField))
+    } yield toSiderealTarget(id, ra, dec, (pmRa, pmDec))
+
+    result.flatMap(identity)
```

Comment:
`result` is a disjunction `CatalogProblem \/ (CatalogProblem \/ Target.SiderealTarget`). There is no `flatten` on the scalaz disjunction but you can do `flatMap` to extract the inner CatalogProblem if it exists.
PS I bet there is a more elegant way to do this :smile: